src/relax/transform/canonicalize_bindings.cc

```diff
@@ -33,68 +33,198 @@
 namespace tvm {
 namespace relax {
 
-class BindingCanonicalizer : public ExprMutator {
+namespace {
+
+struct CanonicalizationPlan {
+  std::unordered_map<Id, Var, ObjectPtrHash, ObjectPtrEqual> replace_usage;
+  std::unordered_map<Id, Var, ObjectPtrHash, ObjectPtrEqual> replace_binding;
+  std::unordered_set<Id, ObjectPtrHash, ObjectPtrEqual> bindings_to_remove;
+};
+
+/*! \brief Utility class to identify usage location
+ *
+ * Canonicalization of a variable binding may require information from
+ * later in the function.  For example, replacing `dataflow_x = expr`
+ * with `var_x = expr` to avoid a trivial binding of `var_x =
+ * dataflow_x` later in the function.  This utility examines a relax
+ * expression, and plans the changes to be made in a mutation pass.
+ */
+class CanonicalizePlanner : public ExprVisitor {
  public:
-  BindingCanonicalizer() {}
-
-  using ExprMutator::VisitExpr_;
-
-  Expr VisitExpr_(const TupleGetItemNode* tuple_get_item) override {
-    if (auto tuple_var = tuple_get_item->tuple.as<Var>()) {
-      if (auto tuple_value = LookupBinding(tuple_var.value())) {
-        if (auto explicit_tuple = tuple_value.as<TupleNode>()) {
-          CHECK_GE(tuple_get_item->index, 0)
-              << "Tuple " << tuple_value << " is accessed at index " << tuple_get_item->index
-              << ", but negative indices are not supported in this context.";
-          CHECK_LT(tuple_get_item->index, explicit_tuple->fields.size())
-              << "Tuple " << tuple_value << " is accessed at index " << tuple_get_item->index
-              << ", but the tuple size is only " << explicit_tuple->fields.size();
-          return VisitExpr(explicit_tuple->fields[tuple_get_item->index]);
+  static CanonicalizationPlan Collect(const Expr& expr) {
+    CanonicalizePlanner visitor;
+    visitor.VisitExpr(expr);
+
+    CanonicalizationPlan plan;
+
+    std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> handled;
+
+    for (const auto& binding_iter : visitor.trivial_bindings_) {
+      Var bound_var = binding_iter.first;
+      Var bound_to = binding_iter.second;
+
+      while (true) {
+        if (auto it = visitor.trivial_bindings_.find(bound_to);
+            it != visitor.trivial_bindings_.end()) {
+          // This may be a trivial binding into a trivial binding.  In
+          // that case, unwrap the bindings until we find the earliest
+          // non-trivial binding.
+          bound_to = it->second;
+        } else {
+          break;
+        }
+      }
```

Comment:
I think this might work instead and might be clearer:
```c++
for (auto it = visitor.trivial_bindings_.find(bound_to);
    it != visitor.trivial_bindings_.end();
    it = visitor.trivial_bindings_.find(bound_to)) {
  bound_to = it->second;
} 
```
(Or the first clause can be moved to before the loop.)

Matter of preference, perhaps, if you consider it simpler than the above.

Reply:
The same applies to the loop below as well.

Reply:
I've gone back and forth on this one a bit.  On the one hand, the current version has an extra level of nesting, but on the other hand, using `for` would require repeating the `.find` clause.  In a perfect world, we would have `while(init_statement; condition)`, but unfortunately (according to [this stackoverflow post](https://stackoverflow.com/q/59985550)), that variation was left out for sake of simplicity.

Reply:
I made another update to use `Map<Id, Expr>` instead of `std::unordered_map<Id, Expr>`, so that the loops could be rewritten as `while(auto opt = lookup.Get(key))`, avoiding the `while(true)` altogether.

Reply:
Yeah, I checked to see if `while (stmt1; condition)` would work and it doesn't. I'm not against having any `while(true)`, but avoiding a level of nesting is usually a good change.